worker/imgproxy.js

```diff
@@ -0,0 +1,180 @@
+import { createHmac } from 'node:crypto'
+import { extractUrls } from '../lib/md.js'
+
+const imgProxyEnabled = process.env.NODE_ENV === 'production' ||
+  (process.env.NEXT_PUBLIC_IMGPROXY_URL && process.env.IMGPROXY_SALT && process.env.IMGPROXY_KEY)
+
+if (!imgProxyEnabled) {
+  console.warn('IMGPROXY_* env vars not set, imgproxy calls are no-ops now')
+}
+
+const IMGPROXY_URL = process.env.NEXT_PUBLIC_IMGPROXY_URL
+const IMGPROXY_SALT = process.env.IMGPROXY_SALT
+const IMGPROXY_KEY = process.env.IMGPROXY_KEY
+
+const cache = new Map()
+
+const knownPositives = [
+  /\.(jpe?g|png|gif|webp|avif)$/,
+  /^https:\/\/i\.postimg\.cc\//,
+  /^https:\/\/i\.imgflip\.com\//,
+  /^https:\/\/i\.imgur\.com\//,
+  /^https:\/\/pbs\.twimg\.com\//,
+  /^https:\/\/www\.zapread\.com\/i\//,
+  /^https:\/\/substackcdn\.com\/image/,
+]
+const knownNegatives = [
+  /^https:\/\/(twitter\.com|x\.com|nitter\.(net|it|at))\/\w+\/status/,
+  /^https:\/\/postimg\.cc/,
+  /^https:\/\/imgur\.com/,
+  /^https:\/\/youtu\.be/,
+  /^https:\/\/(www\.)?youtube\.com/,
+  /^mailto:/,
+  /^https:\/\/stacker\.news\/items/,
+  /^https:\/\/news\.ycombinator\.com\/(item|user)\?id=/,
+  /^https:\/\/\w+\.substack.com/,
+  /^http:\/\/\w+\.onion/,
+  /^http:\/\/nitter\.priv\.loki/,
+  /^http:\/\/\w+\.b32\.i2p/,
+]
```

Comment:
Might make sense to replace this with a [less vulnerable logic](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS)

Reply:
Going to reprocess all images and collect metrics about URLs: how often they were used, how often errors were encountered, how often the HEAD requests returned true or false, how often the GET request returned true or false.

With these metrics, I will create a better list such that we don't need to make any request for most image services our stackers use. We can also skip requests for hosts where we are sure that they are not images. Currently, we try every URL to make sure we don't get false negatives.

Reply:
Collected metrics. See https://github.com/stackernews/stacker.news/pull/513#issuecomment-1735788877

Reply:
> If we immediately sign URLs for common hosting services, this shouldn't be too much of a problem.

I decided not to do this since that would introduce "partial processing" which breaks frontend assumptions about when items have been processed.